### PR TITLE
Improve support for URIs without hosts

### PR DESCRIPTION
**Pull request by rmkerr**
This change fixes a regression in `System.Uri` involving URIs that have no host. The change primarily impacts implicit file URIs and URIs with unknown schemes.

Behavior will change if both of the following are true:
- The URI contains Unicode characters
- The URI is an implicit file URI, or has an unknown scheme and no host

The change is just to revert #9542 and port a fix for the same issue from .NET Framework.

Fixes: #25632

### CODE REVIEW

File: src/System.Private.Uri/src/System/Uri.cs [hunk 1]
```diff
@@ -2141,10 +2141,16 @@ private unsafe ParsingError PrivateParseMinimal()
                         _flags |= Flags.AuthorityFound;
                         idx += 2;
                     }
+                    // There is no Authority component, save the Path index
+                    // Note: mailto is the only guy who is treated specially, should be not.
```
Comment (stephentoub) [inline src/System.Private.Uri/src/System/Uri.cs]: What does "should be not" mean?

File: src/System.Private.Uri/src/System/Uri.cs [hunk 2]
```diff
@@ -2141,10 +2141,16 @@ private unsafe ParsingError PrivateParseMinimal()
...
                     else if (_syntax.NotAny(UriSyntaxFlags.MailToLikeUri))
                     {
-                        // There is no Authority component, save the Path index
-                        // Note: mailto is the only guy who is treated specially, should be not.
+                        // By now we know the URI has no Authority, so if the URI must be normalized, initialize it without one.
+                        if (_iriParsing && (_flags & Flags.HasUnicode) != 0 && (_flags & Flags.HostUnicodeNormalized) == 0)
+                        {
+                            _string = _string.Substring(0, idx);
```
Comment (stephentoub) [inline src/System.Private.Uri/src/System/Uri.cs]: In what situations do we now incur this extra string allocation?

File: src/System.Private.Uri/src/System/Uri.cs [hunk 3]
```diff
@@ -2153,10 +2159,16 @@ private unsafe ParsingError PrivateParseMinimal()
...
+                // There is no Authority component, save the Path index
+                // mailto is treated specially.
                 else if (_syntax.NotAny(UriSyntaxFlags.MailToLikeUri))
                 {
-                    // There is no Authority component, save the Path index
-                    // mailto is treated specially.
+                    // By now we know the URI has no Authority, so if the URI must be normalized, initialize it without one.
+                    if (_iriParsing && (_flags & Flags.HasUnicode) != 0 && (_flags & Flags.HostUnicodeNormalized) == 0)
+                    {
+                        _string = _string.Substring(0, idx);
```
Comment (stephentoub) [inline src/System.Private.Uri/src/System/Uri.cs]: Same question.

File: src/System.Private.Uri/src/System/Uri.cs (hunk 2 above)
Comment (rmkerr) [reply src/System.Private.Uri/src/System/Uri.cs]: We'll hit this code path only when we have a URI that contains Unicode, has a scheme that is not followed by "//" or "///",  and does not have special host parsing (file paths, mailto). This is the same operation as is performed for all other URIs with Unicode. It's just that we missed cutting out everything but the scheme just in this specific case.

If you have any thoughts on how to make the operation more efficient, I'd definitely be open to that though!

File: src/System.Private.Uri/src/System/Uri.cs (hunk 3 above)
Comment (rmkerr) [reply src/System.Private.Uri/src/System/Uri.cs]: Same situation as above, but occurs when the scheme is followed by "///" to specify an empty host.

File: src/System.Private.Uri/src/System/Uri.cs (hunk 1 above)
Comment (rmkerr) [reply src/System.Private.Uri/src/System/Uri.cs]: It means I moved the comment without paying enough attention to what it said. Fixing this now 😄

**Comment by rmkerr** [comment]
@dotnet-bot test this please

**Comment by rmkerr** [comment]
Ubuntu 16.04 on arm didn't get re-run for some reason.
@dotnet-bot test Ubuntu16.04 arm64 Debug

**Comment by rmkerr** [comment]
Pinging @dotnet/ncl for review. One of the CI builds refuses to re-run, but I don't think that the issue is related to this change.

Trying to run it again:
@dotnet-bot test innerloop Ubuntu16.04 arm64 Debug

**Comment by wfurt** [comment]
looks ok to me. arm64 is brand new so it is something to consider. Likely not related to this change.

File: src/System.Private.Uri/src/System/Uri.cs (hunk 2 above)
Comment (rmkerr) [reply src/System.Private.Uri/src/System/Uri.cs]: @stephentoub are you satisfied with the expected performance impact here? If so I'm going to go ahead and merge this.

Comment (stephentoub) [reply src/System.Private.Uri/src/System/Uri.cs]: Ok, thanks.

**Pull request merged**